src/subprocess-posix.cc

```diff
@@ -169,11 +170,61 @@ void SubprocessSet::HandlePendingInterruption() {
     interrupted_ = SIGTERM;
 }
 
+bool SubprocessSet::child_exited_;
+
+void SubprocessSet::SetChildExited(int signum) {
+  child_exited_ = true;
+}
+
+void SubprocessSet::HandleChildExit() {
+  if (!child_exited_)
+    return;
+  child_exited_ = false;
+  for (int i = running_.size(); i > 0; --i)
```

Comment:
You are iterating on running_ in the outer loop, which is strange, as you are also doing waitpid(-1) and iterating on running_ after that in the inner function. I think it should be a 'while waipid() has not reported no-mo-processes' conditioned loop instead.